### ISSUE TITLE
Helper binary for SDKs and CLI

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#3653]
## Overview

User-facing clients (SDKs, CLI, and unofficial clients) need a local binary to enable functionality like local dir sync that has not yet been implemented natively in each client.

## Function

This helper is a low-level tool, not meant to be visible to end users. It implements the complicated process of opening a session to a Dagger engine, so that higher-level clients don't have to. It is used by the official SDKs and CLI published by Dagger, and can also be used by the community to develop unofficial Dagger clients.

The exact features of the helper are still under design. Here is a current snapshot:

| Feature | Description | Status |
| -- | -- | -- |
| Local GraphQL endpoint | Listen for GraphQL queries on a local port, without the need for custom authentication or websocket calls. This keeps clients simple, and allows native compatibility with any graphql tool without modification | *Implemented* |
| Filesync service | Implement the filesync protocol necessary to transfer files back and forth between the client and engine | *Implemented * |
| Interactive terminal service | Implement the protocol necessary to remotely attach to interactive terminals exposed by the engine | *Planned* |
| Socket forwarding | Implement the protocol necessary to forward network traffic back and forth between the client and engine | *Planned* |
| Namespacing | Transparently support namespacing, so that multiple projects can co-exist on the same engine without conflicts | *Planned* |
| Multi-engine | Transparently support connecting to a different engine depending on the client configuration | *Planned* |
| Engine auto-install | Support auto-installing an engine so that everything "just works" | *Proposed* |

## Expected lifetime

Is this helper a temporary stopgap, soon to be deprecated, or a permanent fixture of the Dagger architecture?

* The helper is optional: it's always possible for a client to implement everything itself from scratch.
* It is *recommended* that a new client start by using the helper, for two reasons: 1) it will accelerate development, and 2) it will insulate from the [rapid pace of change in the engine architecture and API](https://github.com/dagger/dagger/issues/3624).
* As a given client becomes more mature, and the underlying architecture of the engine stabilizes, that client *may* choose to stop using the helper.
* The helper itself is considered a *permanent* feature of the Dagger architecture: as more mature clients stop using it, new clients may emerge that will benefit from the speed and stability benefits of the helper.

## Binary name

We must choose a name for the helper binary. Below are some options, a decision has not yet been made.

Notes from Erik:

> this will just be hidden away in XDG_CACHE_HOME in the current preferred approach, so I have very loose opinions here. Any descriptive name that makes it sound like something a human shouldn't invoke directly is good with me.
We should have a help message that tells humans what it is and that they shouldn't invoke it unless they are a robot

| Possible name | Pros | Cons | Current preference (move your name at will) |
| -- | -- | -- | -- |
| `dagger-engine-helper` | generic across everything it might do as we migrate functionality from it to the remote engine | generic name means that function is not intuitively clear | @sipsma  |
| `dagger-engine-init` | | |
| `dagger-engine-connect` | | |
| `dagger-engine-session` | descriptive name that matches the function: a tool that opens a session to the dagger engine | descriptive for the client part of the tool, but not the soon-to-be-migrated server part (API router). The sooner we migrate engine, the less of a problem this is. | @shykes  |

## Binary size

The current helper binary is ~20 MB for linux and darwin on arm64. So ~20 MB x {x86, aarch64} x {linux,darwin,windows} = 120MB total uncompressed size if included in a single package.

## Binary distribution

We need to get this binary to our users, and make sure it matches their OS and architecture.

The preferred option is to **provision the helper binary from the engine OCI image**.

* We need to distribute a container image for our server-side components (mainly buildkitd right now, but more in the future as functionality migrates there).
* We also need each of our SDKs to pull and provision that image anyways.

So we can just put the helper binary in that image and have our SDKs extract it out, installing it in XDG_CACHE_DIR, and using it from there.

This has been prototyped successfully so far here: https://github.com/dagger/dagger/pull/3647

## Command-line syntax

Further subcommand design isn't really worth discussing here since it's 100% internal, can bikeshed here: https://github.com/dagger/dagger/pull/3647

## Additional resources

[Draft implementation](https://github.com/dagger/dagger/pull/3647)

### Other distribution options

Notes from @sipsma when investigating distribution and packaging options:

> | Option | Pros | Cons |
> | -- | -- | -- |
> | Package the binary with each language's native package manager | Least surprising for users, at least for npm/pypi. They just use their package manager and get the dependency they need |  Different in each language and seems to consistently be a can of worms in each language;  Lots of throw-away work if we move to a fully remote engine in future |
> |  Have users install the binary as a pre-requisite | Uses normal well established distribution mechanisms (though `curl \| sh` is frowned upon).  |  Trying to cover every distro and environment can be a massive undertaking; Adds another step to our install instructions, not "magical" |
> | Provision the binary from the engine image |  Re-uses our existing distribution mechanism (engine container image); Easier to optimize pulling of only required binary (don't have to have one fat image that has everything); Clear path to future improvements we want to make to our architecture e.g. next step may be to move https://github.com/dagger/dagger/issues/3629, which should allow the helper binary to stay in the remote engine image and be communicated with over docker exec; Least one-way-door-y; easy to distribute binaries in packages later if we go this route (harder to undo distribution of binaries) | SDKs grab a binary "out-of-band" and uses it on the fly, which raises eyebrows+concerns; Counterpoint: we already pull and execute a container image on the fly which has the ability to import+export your local directories; grabbing a binary out of that image and executing it locally doesn't really expand the scope of what we do that much; Additionally, we'll take steps like checksumming binaries to minimize possible concerns; Some extra code needed in each SDK to handle downloading the helper binary and caching it locally |

Additional notes on native package managers:

> * NPM
>   * Possible but not totally easy to implement: https://blog.xendit.engineer/how-we-repurposed-npm-to-publish-and-distribute-our-go-binaries-for-internal-cli-23981b80911b
>   * requires including binaries for each os+arch combo in the package downloaded by everyone and using custom post-install scripts to set them up
> * PyPi
>   * possible but also not free
>   * Same requirement as npm around including binaries for each os+arch combo.
>   * Default compressed size limit of 60MB (but can be raised if requested)
> * Go
>   * I guess we'd have to commit binaries to our git repo for each os+arch combo and then embed them in our sdk
>   * Huge bloat for user's end binaries
>   * It's not as needed in Go but splitting out the engine from the SDK is very nice to have because it lets us get rid of the replace directives

**Comment by sipsma** [comment]
cc @shykes @aluzzardi 

cc @helderco summary of the discussion around all this and reasons why we're currently pursuing the option in https://github.com/dagger/dagger/pull/3647 (though any and all feedback is still welcome, nothing set in stone)

**Comment by shykes** [comment]
@sipsma can we reduce the scope of this issue to the helper binary? That's already plenty to grok and talk about, and we're lucky that they can now be decoupled.

**Comment by sipsma** [comment]
>  I realize that's kind of already the case, you only mention the user-facing CLI but without going in detail. I'll update the title for clarity.

Thanks, yeah sorry for the confusion. I brought up the user facing CLI because technically it would be possible to make these binaries all the same (one binary for everything) and there is slight interaction in that the CLI itself needs an SDK, but obviously the conclusion here was to decouple the problems as much as possible, which is why I only briefly mention it

**Comment by shykes** [comment]
> > I realize that's kind of already the case, you only mention the user-facing CLI but without going in detail. I'll update the title for clarity.
> 
> Thanks, yeah sorry for the confusion. I brought up the user facing CLI because technically it would be possible to make these binaries all the same (one binary for everything)

That was actually my preference initially :) Since the CLI itself will need all the same features of this helper. But wouldn't make much sense with the way we want to distribute it, I think?

**Comment by shykes** [comment]
> ## (Current Consensus) Option 3 - Provision the binary from the engine image

For the record I agree with the current consensus :)

**Comment by shykes** [comment]
@sipsma will the helper also handle provisioning of the runner (buildkit) image?

**Comment by sipsma** [comment]
> will the helper also handle provisioning of the runner (buildkit) container?

This is still a pretty open question, have gone back and forth, but my current updates to my PR have been moving the provisioning to be in the SDKs. The SDKs then tell the helper exactly where to connect to the remote engine, so the helper is as simple as possible.

My first prototype actually kept the provisioning in the helper and had SDKs only shell out to docker to get the helper binary. However, this approach was complicated (you have to look in both SDK and helper code to know what's going to happen) and felt redundant since the SDKs need to make a `docker create` or a `docker run --rm` call to obtain the binary anyways (cause you can't do `docker cp` on an image ref).

Pros:
1. The code has been easier to follow after making the switch.
1. Eventually we'd most likely want to be able to skip downloading the local binary entirely, so we probably have to go this route anyways eventually.

Cons:
1. Need to ensure each SDK provisions with the same args to `docker run`
   * Note that we already need them to have degrees of compatibility in order to download the helper binaries without stomping over each other, so this only expands the needs for compatibility testing a little more
1. In the long run, we almost certainly can't port every provisioner we'll ever want to every SDK as native code.
   * However, I also don't want to have an eternal dependency on a local binary. So I think we'll probably need to consider (admittedly complicated) "two-phase provisioning" where each SDK has native implementations of simple provisioners (like what I'm writing now) but can get access to more advanced provisioners by bootstrapping the engine with the simple provisioner first.
      * Maybe. I don't love anything about what I just typed there but it's the only viable path to avoiding a hard dep on a local binary forever that I can think of. I think there's some precedence for this sort of bootstrapping but it is definitely complicated to think about.

**Comment by sipsma** [comment]
Getting into more implementation details now: how do we actually connect SDKs with the local engine? The stdio conn approach had a few problems, just got a new solution just barely working, now we need to decide whether we want to spend the time to fix it up now or just fallback to tcp+localhost and switch to the better one in the next release.

---

## The Problem
1. Currently, each `client` needs a single helper/router instance to talk to. This is due to current constraints around session state and lack of multitenancy
2. However, the client itself sometimes sends requests in parallel. We thus want to actually serve those requests in parallel, not just for performance but also because sometimes one request can rely on another finishing (this happens in our test cases sometimes actually)
3. Additionally, we don't always have the luxury of being able to open arbitrary numbers of connections between the SDK and the helper
   * For instance, we both now and in the future need to use stdin/stdout as our connection. But you can't inherently open up multiple streams, it's just one connection, multiplexing has to be layered on top ourselves.
   * We also can't keep spawning new subprocesses due to `#1` above.
4. Ideally, we'd also like this to work on linux, darwin and windows (since clients can run on any of those)

The result of the above is that we need to allow our SDK's graphql clients multiplex graphql requests/responses over a single connection with the router.

## Options

### A) TCP Localhost or Unix Sockets
One way is to have the helper actually bind and listen to a real local endpoint, which could be on localhost or a unix socket (whether on the filesystem or an [abstract socket](https://man7.org/linux/man-pages/man7/unix.7.html)).

Pros:
* Very simple
* Works on linux, darwin and windows (yes, windows actually does have support for unix sockets: https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/)

Cons:
* Only works for local helper binary. When we move to remote engine, this will be thrown away
* Pollutes client's localhost or creates slightly annoying problems around ensuring bound unix sockets get cleaned up in all cases

### B) HTTP 2 over stdio/socketpair

HTTP2 supports multiplexing over a single connection. We can thus use it over stdio pipes (or a [socketpair](https://man7.org/linux/man-pages/man2/socketpair.2.html)) to send graphql requests/responses.

Pros:
* HTTP2 clients are pretty universal across languages (obviously supported on each OS too)
* Works when we want a remote engine in the future

Cons:
* Configuring an HTTP2 client to always only use a single connection seems easier said than done. I could not figure out a guaranteed way to do it w/ Go. There is a connection pool object but in the http2 transport it doesn't seem like it can really be customized to force there to only ever be one connection.
   * That being said, I did get a prototype of this working w/ http2 over stdio pipes, but I was always returning the same connection for the Dial. I didn't break it, but I also didn't see any guarantee that the client would still work even though what it thinks are separate connections are actually just 1. Obscure cases like short writes (i.e. a req/resp is bigger than the socket buffer size) could cause rare issues.
   * I looked quickly at Python but didn't find any obvious way to ensure this either. But I didn't look very much, very low confidence.

### C) Websockets over stdio/socketpair

Websockets inherently support multiplexing due to framing builtin to the protocol. We can send graphql requests w/ an attached ID (random) as a websocket message, have the server read that, run the gql request, and then send a message with the response and the same ID attached.

I got this just *barely* working enough to feel confident it's an option here: https://github.com/sipsma/dagger/commit/08d4f3bf44be3585d3ed6043270df25ed926ab62
* When I say *barely* I mean that I can run test cases successfully, seeing all the req/resp going back and forth, right up until the point where they are shutting down and then I get an error because I didn't include code to handle websocket shutdown messages 🙂
* [client side impl](https://github.com/sipsma/dagger/blob/08d4f3bf44be3585d3ed6043270df25ed926ab62/sdk/go/internal/engineconn/docker/docker.go#L132-L294), [server side impl](https://github.com/sipsma/dagger/blob/08d4f3bf44be3585d3ed6043270df25ed926ab62/cmd/cloak/dial_stdio.go#L72-L142)
* That implementation uses a socketpair because I was getting weird deadlocks when using commandconn w/ stdio pipes. I'm sure there was just a bug somewhere and stdio pipes could be usable, socketpair was just easier for now.

Pros:
* Websocket support is universal across languages
* Easy to control and understand because it's our own little protocol (attaching random ids to requests and waiting for corresponding response), but the protocol is as simple as it gets. The multiplexing/framing complication happens automatically for us in side the websocket libraries
* Websockets may be a transport we want to use in the long run anyways
* Works when we want a remote engine in the future

Cons:
* Somewhat more annoying to add support for in SDKs. We don't need to reimplement websockets, but we do need to be able to plug this custom implementation into graphql clients. Worked in Go pretty easily, from a quick glance it seems like it should be possible in `gql` in Python too.
* Needs more time to be cleaned up and made more robust (not sure yet how much). More thorough testing needed than if we just used off the shelf http2 or something.

## Proposed Solution
I think we should
1. Start in the immediate term with localhost+tcp
2. Once we get all the other engine+sdk release things solidified, if there's time we work on getting the websocket approach fixed up. If we make it for the next release, cool, if we don't, we just switch to that in the next release.

I think this makes the most sense given time constraints. Localhost+tcp's main downside is that it's a throwaway solution, but it's also incredibly simple and robust. It does pollute the host localhost, however we can partially alleviate that with a very simple modification (just bind to a random port, tell the sdk which one to use).

So there will be a little throwaway work up front followed by work to get the better solution, websockets, up to snuff.

---

Going to summarize this and the provisioner approach in a subsequent comment. Just writing this out in detail to get any feedback cc @aluzzardi

**Comment by helderco** [comment]
@sipsma There's also a main upside to localhost+tcp even though it's throwaway. It's what we already have in Python and NodeJS. Anything else would require writing code to support. Even if you got all this working today, I wouldn't have the bandwidth to adjust. What I need is the quickest solution to get that binary so I can keep the rest of the code (for next week's release).

**Comment by sipsma** [comment]
# Summary of what we need from each SDK

cc @helderco 

## Provisioner
1. SDKs will be hardcoded with an image repo + SHA of an engine release that they currently depend on
   * Upgrading the dependency just means changing that const
1. When an SDK needs an engine it first calls:
   * `docker run --name dagger-engine-<sha> -v dagger-engine-<sha> -d --restart always --privileged <image ref + sha> --debug`
   * If it fails with an error containing `Conflict. The container name "/%s" is already in use by container`, then the error can be ignored.
1. Then, the SDK gets `$XDG_CACHE_HOME` (or defaults to `~/.cache`) and checks to see if `$XDG_CACHE_HOME/dagger/dagger-sdk-helper-<sha>` exists. If it does, use it.
   * If it doesn't create a temporary file with a random name under `$XDG_CACHE_HOME/dagger/`. Then run `docker cp dagger-engine-<sha>:/usr/bin/cloak-<os> <name of tmpfile>`, where `<os>` is currently either `linux` or `darwin` depending on the client's os. Windows is TODO.
   * Then do a `rename` on that temp file to the final `$XDG_CACHE_HOME/dagger/dagger-sdk-helper-<sha>`  and use that.
1. Finally, check to see if there are any docker containers with the name prefix `dagger-engine` but with a different SHA, forcibly remove them if so. Do the same for any other helper binaries with different shas under `$XDG_CACHE_HOME/dagger/`
    * The reasons for this are described here (and the previous comment) https://github.com/dagger/dagger/pull/3647#issuecomment-1302721770

## Connecting To Engine
1. Invoke the helper binary provisioned in the previous step with
   * the arg `--remote docker-container://<name of container provisioned with docker run previously>`
   * stderr set to a pipe that is forwarded to the sdk log output stream
2. Wait for the helper binary to write a port number + `\n` to its stdout. This is the random ephemeral port the helper will be listening on
3. Connect to the helper at localhost:<port>, use normal graphql http client from there

**Comment by shykes** [comment]
Last week we had a live discussion with @sipsma about this topic, I want to summarize a few outcomes:

1. The short-term architecture (what we can realistically ship in time for the Python SDK to use) is not exactly what I would like.
2. The difference is not huge between the current architecture and what I would like to see. But it's not worth delaying the Python SDK over.
3. However I still want us to address the difference and resolve it - either by changing the plan to match my preference, convincing me to change my preference, or a mix of both :)

First: @sipsma does this match your recollection?

Assuming the answer is yes, and we agree *my feedback is not a blocker for Python SDK 0.1.0*, here's my main issue with the current architecture: **it does not sufficiently separate 1) helper installation, and 2) engine provisioning.**

I understand that the way the helper is distributed (hitching a ride on the engine image) makes it seem natural to mix those two things into a single logical unit: you already have a dependency on docker, you already used it to fetch the engine image, you already have to run the container (apparently) to access the binary inside it... By that point you've already done almost all the work of our *current* provisioner, why bother separating it out further?

The problem is that the apparent similarity between these two tasks (install the helper; provision the engine) is superficial and temporary:

| | Helper | Engine |
| -- | -- | -- |
| Always auto-installed by the client | Yes | No |
| Only one correct way to install | Yes | No |
| Dependency on Docker is temporary and to be hidden asap | Yes | No |
| Future support for podman, kubernetes, etc. | No | Yes |
| Client in full control of version | Yes | No |
| Possibly affected by config files on the client machine | Yes | No |

At the very least, this difference should be made clear to SDK developers, even if nothing changes in the code they need to write right now. Otherwise, they risk being blindsided later, and forced to refactor their code more painfully than if they were able to plan in advance.

**Comment by shykes** [comment]
New section: "expected lifetime" to clarify whether the helper is a stopgap or more permanent.

**Comment by shykes** [comment]
> @sipsma There's also a main upside to localhost+tcp even though it's throwaway. It's what we already have in Python and NodeJS. Anything else would require writing code to support. Even if you got all this working today, I wouldn't have the bandwidth to adjust. What I need is the quickest solution to get that binary so I can keep the rest of the code (for next week's release).

I don't actually think of localhost+tcp as a throwaway. I think of it as a feature: a simple GraphQL endpoint which any graphql client can query locally, without custom auth or protocols. I guess adding a unix socket option might be nice, to avoid port conflicts.

Then later, we could add http2+stdio which is not as standard, but still pretty standard. But I would rate it as low-priority unless users start asking for it urgently for some unforeseen reason.

I think exposing a custom websocket protocol from the helper does not make sense, since the purpose of the helper is to avoid implementing complicated custom things in the first place. If it's worth the trouble for a client to implement our future websocket-based protocol in full (filesync, terminal attach etc) then they probably don't need the helper anymore.

**Comment by sipsma** [comment]
@shykes Yeah I agree that the differences come down to whether we consider the local helper binary a temporary crutch or a feature. They are two highly divergent futures even if they look similar right now.

Previously, it was very much a temporary crutch in my mind. I still think the simplest possible architecture is that SDKs are fully self-reliant and just connect to a remote engine. But I agree that it's not guaranteed we can move fully remote in the near future (not impossible but not likely). And that if we add more complication to provisioning, it will be easier to handle for now if the SDKs do less of it.

Switching from the current approach will result in one less shelling out to `docker` in SDKs and one more in the engine. So it's a very small lift for now and no immediate harm done, with a chance of upside in the future. I think we should be able to update before the python 0.1 release with this. Worst case scenario and somehow we can't, it's backwards compatible to modify it in a subsequent release.

---

In terms of the naming and the long term intention of what is currently called the helper binary: I'm good with `dagger-engine-session`. It seems to be generic enough to encapsulate both current and future architectures while not being so generic that it could encapsulate literally any architecture (aka "helper").

I did briefly go down a separate lane of thought for a bit that I'll share even though I think it's probably not quite the right approach: if we are going to truly embrace the use of a local binary rather than considering it a short term crutch, we could consider calling it the `dagger-local-engine`. Conceptually this will mean:
1. The local engine is the binary that SDKs communicate with to run pipelines.
2. The local engine is not *always* self sufficient though and may instead relay requests to a "remote engine" where queries are actually processed and executed.

In our architecture at this exact moment, the above description is accurate. But even in the longer term when we move router to be remote, you could argue that the local engine has just offloaded even more of its capabilities to the remote engine.

However, I think that last part is a pretty tenuous argument, which is why I'm good with `dagger-engine-session`.

---

> I don't actually think of localhost+tcp as a throwaway. I think of it as a feature: a simple GraphQL endpoint which any graphql client can query locally, without custom auth or protocols. I guess adding a unix socket option might be nice, to avoid port conflicts.

The longer term downside of localhost is that it means anyone with `CAP_NET_ADMIN` can listen to the communications, so it's not an ideal channel for secrets. Unix sockets are better in this regard. But this isn't really a blocker right now.

> Then later, we could add http2+stdio which is not as standard, but still pretty standard. But I would rate it as low-priority unless users start asking for it urgently for some unforeseen reason.

I wish http2 over stdio pipes was as easy as it should be, but what I discovered is that it requires being able to force the client to only ever multiplex everything over a single channel (or else you will get corner cases where a `write` to the pipe exceeds `PIPE_BUF` and gets interleaved). We shouldn't pursue this unless we're willing to do a lot of testing. Websockets were conceptually easier to deal with which is why I prototyped it quickly (but agree that also only makes sense in the grander scheme of things where we are using websockets as part of a broader api, which is kind of where I was leaning at the time).

**Comment by sipsma** [comment]
We continued this thread here: https://github.com/dagger/dagger/issues/3830
